torch/csrc/jit/passes/lower_tuples.cpp

```diff
@@ -5,18 +5,58 @@
 
 namespace torch { namespace jit {
 
+namespace {
+
 // operators where we expect to find tuples as inputs/outputs
-// this is to assert we are only  doing modifications when we know
+// this is to assert we are only doing modifications when we know
 // we can flatten tuples
 std::unordered_set<Symbol> white_list = {
   prim::If,
   prim::Loop,
   prim::TupleUnpack,
   prim::TupleConstruct,
+  prim::TupleIndex,
+  prim::TupleSlice,
   prim::Param,
   prim::Return,
 };
 
+void removeTupleNodes(Node *n, bool must_remove_tuples) {
+  if (n->kind() != prim::TupleUnpack && n->kind() != prim::TupleIndex
+      && n->kind() != prim::TupleSlice) {
+    return;
+  }
+  auto construct = n->input()->node();
+  if (construct->kind() != prim::TupleConstruct) {
+    if (must_remove_tuples) {
+      AT_ERROR(n->kind().toQualString(), " not matched to tuple construct");
+    }
+    return;
+  }
+  if (n->kind() == prim::TupleUnpack) {
+    for(size_t i = 0; i < n->outputs().size(); ++i) {
+      n->outputs()[i]->replaceAllUsesWith(construct->inputs().at(i));
+    }
+  } else if (n->kind() == prim::TupleIndex) {
+    auto idx = n->i(attr::index);
+    n->output()->replaceAllUsesWith(construct->inputs().at(idx));
+  } else if (n->kind() == prim::TupleSlice) {
+    std::vector<Value*> values;
+    int64_t beg = n->i(attr::beg);
+    int64_t end = n->i(attr::end);
+    int64_t step = n->i(attr::step);
+    for (size_t i = beg; i < end; i += step) {
+      values.push_back(construct->inputs().at(i));
+    }
+    auto graph = n->owningGraph();
+    auto tuple_out = graph->createTuple(values);
+    graph->setInsertPoint(n);
```

Comment:
Use `WithInsertGuard`. Otherwise you are globally setting the insert point of the graph. This will lead to subtle bugs.